### PR TITLE
Do not hold timelines lock during GC

### PR DESCRIPTION
**Pull request by knizhnik**
refer #1087

### CODE REVIEW

**Comment by hlinnaka** [comment]
This doesn't seem safe. See the piece of code where we acquire the lock:
```
        // grab mutex to prevent new timelines from being created here.
        // TODO: We will hold it for a long time
        let mut timelines = self.timelines.lock().unwrap();
```
If we drop the lock, then a branch could be created that depends on the old page versions we're just about to garbage-collect away. 

Maybe there's some other mechanism that prevents that. Maybe we can be sure that you can only create a branch within the GC horizon, and we never garbage collect stuff newer than the GC horizon? I don't think we have any hard guarantee on that either at the moment though. At least this needs comments to explain why it's safe.

**Comment by knizhnik** [comment]
> This doesn't seem safe. See the piece of code where we acquire the lock:
> 
> ```
>         // grab mutex to prevent new timelines from being created here.
>         // TODO: We will hold it for a long time
>         let mut timelines = self.timelines.lock().unwrap();
> ```
> 
> If we drop the lock, then a branch could be created that depends on the old page versions we're just about to garbage-collect away.
> 
> Maybe there's some other mechanism that prevents that. Maybe we can be sure that you can only create a branch within the GC horizon, and we never garbage collect stuff newer than the GC horizon? I don't think we have any hard guarantee on that either at the moment though. At least this needs comments to explain why it's safe.

It seems to be that the simplest solution will be just to add separate mutex for GC: it will prevetn creation of new timelines during GC, but doesn't block get_timeline and get_timeline_status.

Concerning your proposal to prohibit creation creation of branches behind GC cutoff - it may be really better behavior, because provide more predictable and clear behavior for users. Now it may be possible to create branch behind GC horizon if GC was not activated since the moment of crossing this boundary. As far as GC may be started quite rarely, then in most cases user will be able to create branches in the past, even with zero PiTR interval. But sometimes such request may fail with `invalid branch start lsn` error.  It may be safer it use will always get this error if branch start lsn is behind GC horizon, doesn't matter whether GC was actually performed or not.

But such check also requires critical section, because if  `start branch lsn` is ok at this moment, it may become invalid a moment later.

So I just wait for your comment concerning introduction of `gc_cs` mutex. If you agree, I will add more comments explaining need in yet another mutex.

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -127,7 +127,7 @@ pub struct LayeredRepository {
     conf: &'static PageServerConf,
     tenantid: ZTenantId,
     timelines: Mutex<HashMap<ZTimelineId, LayeredTimelineEntry>>,
-
+    gc_cs: Mutex<i32>,
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:136]:
```suggestion
    gc_cs: Mutex<()>,
```

Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:136]: What does 'cs' stand for?

**Review by hlinnaka — APPROVED** [review]

**Comment by hlinnaka** [comment]
> It seems to be that the simplest solution will be just to add separate mutex for GC: it will prevetn creation of new timelines during GC, but doesn't block get_timeline and get_timeline_status.

+1. We might want to do something smarter in the future to allow creating branches while GC is running, but that will do for now.

**Comment by LizardWizzard** [comment]
> Concerning your proposal to prohibit creation creation of branches behind GC cutoff - it may be really better behavior

It is how it works in `check_lsn_is_in_scope`. The problem is if there is no mutex the race condition is possible. I agree that separate mutex is the easiest way to solve it for now.

> doesn't matter whether GC was actually performed or not

This might be solved by advancing gc horizon even without actual gc running. But this may be harder to tweak/reason about (introduces advance interval separately of gc interval) so may be not a good fit.

File: pageserver/src/layered_repository.rs (hunk 1 above)
Comment (knizhnik) [reply pageserver/src/layered_repository.rs]: critical section

**Comment by knizhnik** [comment]
> > Concerning your proposal to prohibit creation creation of branches behind GC cutoff - it may be really better behavior
> 
> It is how it works in `check_lsn_is_in_scope`. The problem is if there is no mutex the race condition is possible. I agree that separate mutex is the easiest way to solve it for now.

`check_lsn_is_in_scope` is looking at `latest_gc_cutoff_lsn` which is advanced by GC.
If GC was not activated, this variable is not updated too.


> 
> > doesn't matter whether GC was actually performed or not
> 
> This might be solved by advancing gc horizon even without actual gc running. But this may be harder to tweak/reason about (introduces advance interval separately of gc interval) so may be not a good fit.

Not clear who and when should do it.
We can just subtract `gc_horizon' from  `last_record_lsn` and to compare with branch start point. But it can be a source of race condition if not protected by mutex.

**Pull request merged**